File: App/Resources/AlgorithmDetails/optimizedlazystablesort/optimizedlazystablesort.cpp
#include <cstdio>
#include <utility>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

void swapAt(int arr[], int a, int b) { std::swap(arr[a], arr[b]); }

void multiSwap(int arr[], int a, int b, int count) {
  for (int i = 0; i < count; i++)
    swapAt(arr, a + i, b + i);
}

void rotate(int arr[], int pos, int lenA, int lenB) {
  while (lenA != 0 && lenB != 0) {
    if (lenA <= lenB) {
      multiSwap(arr, pos, pos + lenA, lenA);
      pos += lenA;
      lenB -= lenA;
    } else {
      multiSwap(arr, pos + (lenA - lenB), pos + lenA, lenB);
      lenA -= lenB;
    }
  }
}

int binSearch(int arr[], int pos, int len, int keyPos, int isLeft) {
  int left = -1;
  int right = len;
  int key = arr[keyPos];
  while (left < right - 1) {
    int mid = left + (right - left) / 2;
    int cond = isLeft ? (arr[pos + mid] >= key) : (arr[pos + mid] > key);
    if (cond)
      right = mid;
    else
      left = mid;
  }
  return right;
}

void mergeWithoutBuffer(int arr[], int pos, int len1, int len2) {
  if (len1 == 0 || len2 == 0)
    return;
  if (len1 < len2) {
    while (len1 != 0) {
      int loc = binSearch(arr, pos + len1, len2, pos, 1);
      if (loc != 0) {
        rotate(arr, pos, len1, loc);
        pos += loc;
        len2 -= loc;
      }
      if (len2 == 0)
        break;
      do {
        pos++;
        len1--;
      } while (len1 != 0 && arr[pos] <= arr[pos + len1]);
    }
  } else {
    while (len2 != 0) {
      int loc = binSearch(arr, pos, len1, pos + len1 + len2 - 1, 0);
      if (loc != len1) {
        rotate(arr, pos + loc, len1 - loc, len2);
        len1 = loc;
      }
      if (len1 == 0)
        break;
      do {
        len2--;
      } while (len2 != 0 && arr[pos + len1 - 1] <= arr[pos + len1 + len2 - 1]);
    }
  }
}

/* Guard: a chunk of length <= 1 has nothing to compare. ArrayV's own source
   skips this check and unconditionally reads arr[a] / arr[a + 1], which crashes
   whenever chunking leaves a trailing 1-element chunk (e.g. n = 17 leaves a
   final [16, 17) chunk). */
void insertionSortChunk(int arr[], int a, int b) {
  if (b - a <= 1)
    return;
  int i = a + 1;
  bool descending = arr[i - 1] > arr[i];
  i++;
  if (descending) {
    while (i < b && arr[i - 1] > arr[i])
      i++;
    int lo = a, hi = i - 1;
    while (lo < hi) {
      swapAt(arr, lo, hi);
      lo++;
      hi--;
    }
  } else {
    while (i < b && arr[i - 1] <= arr[i])
      i++;
  }
  while (i < b) {
    int current = arr[i];
    int pos = i - 1;
    while (pos >= a && arr[pos] > current) {
      arr[pos + 1] = arr[pos];
      pos--;
    }
    arr[pos + 1] = current;
    i++;
  }
}

void lazyStableSort(int arr[], int pos, int len) {
  int dist = 0;
  while (dist + 16 < len) {
    insertionSortChunk(arr, pos + dist, pos + dist + 16);
    dist += 16;
  }
  if (dist < len)
    insertionSortChunk(arr, pos + dist, pos + len);

  int part = 16;
  while (part < len) {
    int left = 0;
    int right = len - 2 * part;
    while (left <= right) {
      mergeWithoutBuffer(arr, pos + left, part, part);
      left += 2 * part;
    }
    int rest = len - left;
    if (rest > part)
      mergeWithoutBuffer(arr, pos + left, part, rest - part);
    part *= 2;
  }
}

void sort(int arr[], int n) { lazyStableSort(arr, 0, n); }

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: Modules/BuiltInAlgorithms/Sources/Merge/OptimizedLazyStableSort.swift
import AlgorithmKit
import SortEngineKit

/// Ported from ArrayV's `OptimizedLazyStableSort` — overrides `GrailSortingTemplate`'s own
/// `lazyStableSort` entirely with a different construction (natural-run-detecting insertion sort
/// over fixed 16-element chunks, then doubling `mergeWithoutBuffer`), while still reusing the
/// template's `mergeWithoutBuffer` unmodified. ArrayV's own subclass genuinely does this (a real
/// override, not decoration) — see `TEMPLATE_PORT_REFERENCE.md` §6.
///
/// **Stability**: `insertionSort`'s shifts use `engine.setValue`, so the standard swap-tape-shadow
/// stability test can't observe them and produces false failures if pointed at this algorithm
/// (confirmed: 50/50 spurious "failures"). Verified genuinely stable instead by simulating this
/// exact algorithm in Python with a parallel original-index array threaded through every swap and
/// write (2,000 randomized duplicate-heavy trials, zero wrong results, zero instability) — the
/// natural-run detection only reverses strictly-decreasing runs (no ties inside by definition),
/// the plain insertion-sort continuation only shifts strictly-greater elements, and
/// `mergeWithoutBuffer` is already independently confirmed stable via `GrailSort`'s own test.
public struct OptimizedLazyStableSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "optimizedlazystablesort")
  public let metadata = AlgorithmMetadata(
    displayName: "Optimized Lazy Stable Sort",
    category: .merge,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 462, coefficients: [239433, 1020.51, 1.08469],
      measuredSafeCeiling: nil),
    stable: true,
    timeComplexity: ComplexityBounds(
      best: "O(n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(1)",
    iconName: "hare"
  )

  public init() {}

  /// Detects a single leading natural run (ascending-or-equal is scanned as-is; a strictly
  /// descending run is scanned then reversed in place), then falls through to a standard
  /// insertion sort for whatever's left — reduces comparisons on already-mostly-ordered chunks
  /// versus a plain insertion sort.
  ///
  /// **Real bug found and fixed**: ArrayV's own `insertionSort(array, a, b, ...)` reads
  /// `array[a]`/`array[a+1]` unconditionally before any bounds check — for `n` not a multiple of
  /// 16, the final chunk passed in by `record(into:)` below can be exactly 1 element wide
  /// (confirmed crash: `n = 17` produces a `[16, 17)` tail chunk), which faithfully reproduced
  /// reads one past the valid range. A single-element range is already trivially sorted and needs
  /// no comparison at all, so this guard is a correctness fix, not a behavior change on any input
  /// that previously worked.
  private func insertionSort(_ engine: inout RecordingEngine, _ a: Int, _ b: Int) {
    guard b - a > 1 else { return }
    var i = a + 1
    if engine.compare(i - 1, i, by: >) {
      i += 1
      while i < b && engine.compare(i - 1, i, by: >) { i += 1 }
      engine.reversal(a, i - 1)
    } else {
      i += 1
      while i < b && engine.compare(i - 1, i, by: <=) { i += 1 }
    }

    while i < b {
      let current = engine.values[i]
      var pos = i - 1
      while pos >= a && engine.values[pos] > current {
        engine.setValue(pos + 1, engine.values[pos])
        pos -= 1
      }
      engine.setValue(pos + 1, current)
      i += 1
    }
  }

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n > 1 else { return }

    var dist = 0
    while dist + 16 < n {
      insertionSort(&engine, dist, dist + 16)
      dist += 16
    }
    if dist < n {
      insertionSort(&engine, dist, n)
    }

    var part = 16
    while part < n {
      var left = 0
      let right = n - 2 * part
      while left <= right {
        GrailSortingTemplate.mergeWithoutBuffer(&engine, left, part, part)
        left += 2 * part
      }
      let rest = n - left
      if rest > part {
        GrailSortingTemplate.mergeWithoutBuffer(&engine, left, part, rest - part)
      }
      part *= 2
    }
  }
}
